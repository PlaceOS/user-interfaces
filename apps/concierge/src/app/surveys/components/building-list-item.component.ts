import { Component, OnInit, Input } from '@angular/core';
import { Building } from '@placeos/organisation';
import { Router, ActivatedRoute } from '@angular/router';
import { BuildingListItemService } from '../services/building-list-item.service';
import { shareReplay } from 'rxjs/operators';
import { SettingsService } from '@placeos/common';

@Component({
    selector: 'building-list-item',
    template: `
        <section
            class="building-wrapper bg-base-100 my-4 mx-auto rounded-md overflow-hidden"
        >
            <div *ngIf="building.image" class="image-container">
                <img
                    auth
                    class="flex object-fill "
                    [source]="building.image"
                    alt="image of building"
                />
            </div>
            <div *ngIf="!building.image" class="image-container">
                <span> <app-icon class="domain-icon">domain</app-icon></span>
            </div>
            <div class="details-container my-4 mx-8 w-full relative">
                <div class="location-wrapper">
                    <app-icon
                        class="location-icon"
                        aria-hidden="false"
                        aria-label="Material icon of location pointer"
                        >location_on</app-icon
                    >
                    <span> {{ building.address }} </span>
                </div>

                <span class="building-title">
                    {{ building.display_name }}
                </span>
                <ng-container
                    *ngIf="!(loading$ | async)?.length; else loadState"
                >
                    <ng-container *ngIf="stats$ | async as stats">
                        <div class="flex py-4 justify-end space-x-2">
                            <div class="flex flex-col items-center flex-1">
                                <h3>Live Surveys</h3>
                                <p class="text-4xl">
                                    {{ stats?.lives || 0 }}
                                </p>
                            </div>
                            <div class="flex flex-col items-center flex-1">
                                <h3>Draft Surveys</h3>
                                <p class="text-4xl">
                                    {{ stats?.drafts || 0 }}
                                </p>
                            </div>
                            <div class="flex flex-col items-center flex-1">
                                <h3>Total Answers</h3>
                                <p class="text-4xl">
                                    {{ stats?.responses || 0 }}
                                </p>
                            </div>
                        </div>
                    </ng-container>
                </ng-container>
                <div class="flex justify-end w-full mt-4">
                    <button btn matRipple class="inverse" (click)="navigate()">
                        <span class="ml-2">View</span>
                        <app-icon class="ml-1 text-2xl">chevron_right</app-icon>
                    </button>
                </div>
            </div>
        </section>
        <ng-template #loadState>
            <div class="flex absolute inset-0 opacity-60 bg-base-100 z-10">
                <div class="flex flex-col m-auto items-center">
                    <mat-spinner [diameter]="32"></mat-spinner>
                    <span>{{ loading$ | async }}</span>
                </div>
            </div>
        </ng-template>
    `,
    styles: [
        `
            .building-wrapper {
                display: flex;
                position: relative;
                flex-direction: row;
                position: relative;
                justify-content: flex-start;
                height: 260px;
                max-width: 68rem;
                border: 1px solid #e6e6e6;
                box-shadow: 0px 2px 4px rgba(5, 28, 44, 0.1);
            }
            .options {
                display: flex;
                align-items: center;
                justify-content: center;
                background-color: var(--b1);
                position: absolute;
                height: 48px;
                width: 48px;
                top: 10px;
                left: 10px;
                border: 1px solid #e6e6e6;
                border-radius: 3px;
            }
            .menu-wrapper {
                display: flex;
                flex-direction: column;
                height: 90px;
                width: 135px;
                justify-content: center;
            }
            .image-container {
                display: flex;
                height: 100%;
                width: 380px;
                background-color: rgba(0, 0, 0, 0.22);
                align-items: center;
                justify-content: center;
            }
            .details-container {
                display: flex;
                flex-direction: column;
                min-width: 600px;
            }
            .location-wrapper {
                display: flex;
                flex-direction: row;
                align-items: center;
                margin-bottom: 20px;
            }
            .location-icon {
                color: #5295f7;
                margin-right: 10px;
            }
            .domain-icon {
                display: flex;
                color: #fff;
                height: 100%;
                width: 100%;
            }
            .building-title {
                font-size: 18px;
                font-weight: 500;
            }
            .details-text {
                display: flex;
                flex-direction: column;
                font-size: 14px;
                font-weight: 400;
                color: #020202;
                margin-bottom: 10px;
                justify-content: space-between;
                height: 70px;
            }
            .button-container {
                display: flex;
                position: absolute;
                bottom: 20px;
                right: 20px;
            }
            /* .view-button {
                display: inline-flex;
                color: #292f5b;
                background-color: var(--b1);
                border-radius: 2px;
                margin: 20px 0px 20px 20px;
                justify-content: center;
                padding-left: 20px;
            }
            .view-button span {
                display: inline-flex;
            }
            .view-button app-icon {
                display: flex;
                justify-content: center;
                align-items: center;
            } */
        `,
    ],
    providers: [BuildingListItemService],
})
export class BuildingListItemComponent implements OnInit {
    @Input() building: Building | any;

    loading$ = this.service.loading$.pipe(shareReplay(1));
    stats$ = this.service.stats$.pipe(shareReplay(1));

    constructor(
        private _settings: SettingsService,
        private router: Router,
        private service: BuildingListItemService
    ) {}

    ngOnInit(): void {
        this.service.initStats(this.building.id);
    }

    navigate(): void {
        this.router.navigate([
            this._settings.get('app.default_route').includes('new')
                ? '/surveys/new'
                : '/survey',
            'survey-list',
            this.building.id,
        ]);
    }
}
