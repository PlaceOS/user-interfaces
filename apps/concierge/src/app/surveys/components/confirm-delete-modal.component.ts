import { Component, OnInit, Inject } from '@angular/core';
import { MatDialog, MatDialogRef } from '@angular/material/dialog';

@Component({
    selector: 'confirm-delete-modal',
    template: `
        <aside>
            <span
                >Are you sure you want to permanently delete this survey?</span
            >
            <div class="button-wrapper">
                <button
                    mat-button
                    (click)="cancel()"
                    class="cancel-button"
                    aria-label="do not delete survey"
                >
                    No
                </button>
                <button
                    mat-flat-button
                    (click)="delete()"
                    class="delete-button"
                    aria-label="delete survey"
                >
                    Yes Delete
                </button>
            </div>
        </aside>
    `,
    styles: [
        `
            aside {
                margin: 20px;
            }
            .button-wrapper {
                display: flex;
                padding-top: 20px;
                width: 100%;
                justify-content: space-between;
            }
            .cancel-button {
                background-color: #fff;
                border: 1px solid rgba(0, 0, 0, 0.12);
                color: black;
                border-radius: 2px;
                font-weight: 700;
                padding: 6px 20px;
            }
            .delete-button {
                background-color: #e12e40;
                color: #fff;
                font-weight: 700;
                border-radius: 2px;
                padding: 6px 20px;
            }
        `,
    ],
})
export class ConfirmDeleteModalComponent implements OnInit {
    constructor(public dialogRef: MatDialogRef<ConfirmDeleteModalComponent>) {}

    ngOnInit(): void {}

    cancel(): void {
        this.dialogRef.close(false);
    }

    delete() {
        this.dialogRef.close(true);
    }
}
