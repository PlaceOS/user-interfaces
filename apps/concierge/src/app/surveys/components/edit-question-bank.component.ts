import { Component, OnInit, OnDestroy } from '@angular/core';
import { Observable, Subscription, BehaviorSubject, of } from 'rxjs';
import { MatDialog, MatDialogRef } from '@angular/material/dialog';
import { FormControl } from '@angular/forms';
import { Question, QuestionType } from '@placeos/survey-suite';
import { SurveyCreatorService } from '../services/survey-creator.service';
import { SearchService } from '../services/search.service';

@Component({
    selector: 'edit-question-bank',
    template: `
        <section>
            <div class="header">
                <span class="dialog-title">Edit question bank</span>
                <span class="close" (click)="closeDialog()">
                    <mat-icon
                        aria-hidden="false"
                        aria-label="Material icon for closing dialog"
                        class="icon"
                        >close</mat-icon
                    >
                </span>
            </div>
            <div class="search-bar-container">
                <search-bar [view]="'checkbox'"></search-bar>
            </div>
            <main>
                <span class="question-counter">
                    {{ (this.question_bank$ | async).length }} questions
                </span>
                <div *ngFor="let question of this.question_bank$ | async">
                    <question-container
                        (newTitleEvent)="updateTitle($event, question.title)"
                        (newRatingEvent)="updateRating($event, question.title)"
                        (newChoiceEvent)="updateChoice($event, question.title)"
                        (allChoicesEvent)="
                            updateAllChoices($event, question.title)
                        "
                        (updateTypeEvent)="updateType($event, question.title)"
                        [question]="question"
                        [canDelete]="true"
                    ></question-container>
                </div>
            </main>
            <footer>
                <button
                    mat-button
                    class="cancel-button"
                    color="basic"
                    (click)="closeDialog()"
                >
                    Cancel
                </button>
                <button
                    mat-button
                    class="update-button"
                    color="primary"
                    (click)="updateQuestions()"
                >
                    Update
                </button>
            </footer>
        </section>
    `,
    styles: [
        `
            ::ng-deep .mat-select-value {
                max-width: min-content;
                margin-right: 3px;
            }
            .header {
                display: flex;
                align-items: center;
                justify-content: space-between;
                background-color: #fff;
                height: 70px;
                border: 1px solid rgba(0, 0, 0, 0.12);
            }
            main {
                margin: 0px;
                padding: 20px;
                background-color: #f8f8fa;
                border: 1px solid #d4d4d4;
            }
            .dialog-title {
                display: flex;
                font-size: 20px;
                line-height: 30px;
                font-weight: 800;
                padding: 20px;
            }
            .section-heading {
                font-size: 16px;
                font-weight: 700;
                line-height: 24px;
            }
            .small-text {
                size: 11px;
                font-weight: 400;
                margin-top: 3px;
                color: #808080;
            }
            .search-bar-container {
                padding-top: 20px;
            }
            .question-counte {
                font-size: 20px;
                line-height: 24px;
                weight: 400;
            }
            .new-question {
                margin-left: -10px;
            }
            .cancel-button {
                display: flex;
                color: #292f5b;
                background-color: #fff;
                border-radius: 2px;
                margin: 20px 0px 20px 20px;
            }
            .update-button {
                display: flex;
                color: #fff;
                background-color: #292f5b;
                border-radius: 2px;
                margin: 20px 0px 20px 20px;
            }
            footer {
                display: flex;
                flex-direction: row;
                justify-content: flex-end;
                margin-right: 20px;
            }
            .close {
                display: flex;
                z-index: 99;
                text-align: right;
                vertical-align: middle;
                line-height: 12px;
                margin-right: 20px;
            }
        `,
    ],
})
export class EditQuestionBankComponent implements OnInit {
    updated_question_bank: Question[] = [];
    question_bank$: Observable<Question[]>;
    bank_sub: Subscription;
    flag_sub: Subscription;
    private _update_flag: BehaviorSubject<boolean> =
        new BehaviorSubject<boolean>(false);

    public QuestionType = QuestionType;

    constructor(
        public dialogRef: MatDialogRef<EditQuestionBankComponent>,
        public surveyCreatorService: SurveyCreatorService,
        public searchService: SearchService
    ) {}

    ngOnInit(): void {
        this.bank_sub = this.surveyCreatorService.question_bank$.subscribe(
            (questions: Question[]) => (this.updated_question_bank = questions)
        );
        this.question_bank$ = this.searchService.question_bank$;
    }

    closeDialog() {
        this.dialogRef.close();
    }

    updateTitle(event, title) {
        let found_question = this.updated_question_bank.find(
            (question) => question.title === title
        );
        this.flag_sub = this._update_flag.asObservable().subscribe((flag) => {
            if (flag) {
                found_question.title = event[0];
                found_question.type = event[1];
            }
        });
        this.surveyCreatorService.question_bank = this.updated_question_bank;
        this._update_flag.next(false);
    }

    updateChoice(event, title) {
        let found_question = this.updated_question_bank.find(
            (question) => question.title === title
        );
        this.flag_sub = this._update_flag.asObservable().subscribe((flag) => {
            if (flag) {
                let choice_index = found_question.choices.findIndex(
                    (item) => item == event[1]
                );
                found_question.choices[choice_index] = event[0];
            }
        });
        this.surveyCreatorService.question_bank = this.updated_question_bank;
        this._update_flag.next(false);
    }

    updateAllChoices(event, title) {
        let found_question = this.updated_question_bank.find(
            (question) => question.title === title
        );
        this.flag_sub = this._update_flag.asObservable().subscribe((flag) => {
            if (flag) {
                found_question.choices = event;
            }
        });
        this.surveyCreatorService.question_bank = this.updated_question_bank;
        this._update_flag.next(false);
    }
    updateRating(event, title) {
        let found_question = this.updated_question_bank.find(
            (question) => question.title === title
        );
        this.flag_sub = this._update_flag.asObservable().subscribe((flag) => {
            if (flag) {
                found_question.rateValues = event;
            }
        });
        this.surveyCreatorService.question_bank = this.updated_question_bank;
        this._update_flag.next(false);
    }

    updateType(event, title) {
        let found_question = this.updated_question_bank.find(
            (question) => question.title === title
        );

        found_question.type = event;

        if (event == QuestionType.Multi_Select || QuestionType.Single_Select) {
            found_question.choices = ['Type a choice here...'];
        }

        this.surveyCreatorService.question_bank = this.updated_question_bank;
    }

    updateQuestions() {
        this._update_flag.next(true);
        this.closeDialog();
    }

    ngOnDestroy(): void {
        this.bank_sub?.unsubscribe();
        this.flag_sub?.unsubscribe();
    }
}
