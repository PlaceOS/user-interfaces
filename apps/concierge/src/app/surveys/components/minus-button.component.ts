import { Component, OnInit } from '@angular/core';

@Component({
    selector: 'minus-button',
    template: `
        <div class="button-container">
            <button mat-mini-fab aria-label="Button with plus icon">
                <mat-icon>remove</mat-icon>
            </button>
        </div>
    `,
    styles: [
        `
            .button-container {
                display: flex;
                justify-content: center;
                width: 40px;
            }

            .button-container button {
                display: flex;
                position: relative;
                background-color: #9a2d2d;
                height: 20px;
                width: 20px;
                justify-content: center;
            }

            .button-container button mat-icon {
                position: absolute;
                top: 0;
                left: 0;
                margin-left: -2px;
                margin-top: 1px;
                font-size: 18px;
                font-weight: 700;
            }
        `,
    ],
})
export class MinusButtonComponent implements OnInit {
    constructor() {}

    ngOnInit(): void {}
}
