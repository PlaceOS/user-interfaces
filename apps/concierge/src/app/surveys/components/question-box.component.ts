import { Component, OnInit } from '@angular/core';

@Component({
    selector: 'question-box',
    template: ` <div class="box">
        <input-title [placeholder]="'Page 1'" [fontSize]="'24'"></input-title>
        <input-title
            [placeholder]="'Description'"
            [fontSize]="'16'"
        ></input-title>
        <div class="question-wrapper">
            <div class="question-row">
                <span class="question-number">1.</span>
                <input-title
                    [placeholder]="'question1'"
                    [fontSize]="'16'"
                ></input-title>
            </div>
        </div>
        <div class="question-wrapper"></div>
    </div>`,
    styles: [
        `
            .box {
                display: flex;
                flex-direction: column;
                width: 600px;
                background-color: #f5f5f5;
                margin: auto;
                border: 3px solid #808080;
                border-radius: 15px;
            }

            .box:hover,
            .box:active {
                border: 3px solid orange;
                background-color: #ffeed2;
            }

            .question-wrapper {
                background-color: #fff;
                margin: 5px 15px 15px 15px;
                padding: 20px;
                border-radius: 10px;
            }

            .question-number {
                font-size: 12px;
                color: #808080;
            }

            .question-row {
                display: flex;
                flex-direction: row;
                align-items: center;
            }
        `,
    ],
})
export class QuestionBoxComponent implements OnInit {
    constructor() {}

    ngOnInit(): void {}
}
