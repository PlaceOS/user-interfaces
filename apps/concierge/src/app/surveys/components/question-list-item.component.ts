import { Component, OnInit, Input } from '@angular/core';
import { Question } from '../survey-types';

@Component({
    selector: 'question-list-item',
    template: `
        <div class="question-list-item">
            <span class="icon-span">
                <mat-icon
                    aria-hidden="false"
                    aria-label="Material icon for button"
                    class="icon"
                    >drag_indicator</mat-icon
                >
            </span>
            <span>
                {{ question.title }}
            </span>
            <span *ngIf="question.selected">
                <mat-icon
                    aria-hidden="false"
                    aria-label="Material tick icon"
                    class="tick-icon"
                    >done</mat-icon
                >
            </span>
        </div>
    `,
    styles: [
        `
            .question-list-item {
                display: flex;
                position: relative;
                background-color: #fff;
                align-items: center;
                margin: 10px;
                padding: 10px;
                border-radius: 5px;
                border: 1px solid #f5f5f5;
                font-size: 12px;
                color: #808080;
            }
            .question-list-item span {
                display: inline-flex;
                vertical-align: center;
            }
            .icon {
                display: inline-flex;
                vertical-align: center;
                justify-content: center;
                align-items: center;
                margin: 0px 5px 0px 5px;
            }
            .tick-icon {
                position: absolute;
                top: 10px;
                right: 0;
                color: #529a60;
            }
        `,
    ],
})
export class QuestionListItemComponent implements OnInit {
    @Input() question: Question;

    constructor() {}

    ngOnInit(): void {}
}
