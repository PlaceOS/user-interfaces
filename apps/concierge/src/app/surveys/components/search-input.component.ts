import {
    AfterViewInit,
    Component,
    ElementRef,
    EventEmitter,
    Input,
    Output,
    ViewChild,
} from '@angular/core';
import { AsyncHandler } from '@placeos/common';
import { fromEvent, Observable } from 'rxjs';
import { debounceTime, distinctUntilChanged, map } from 'rxjs/operators';

@Component({
    selector: 'search-input',
    styles: [],
    template: `
        <mat-form-field
            appearance="outline"
            subscriptSizing="dynamic"
            class="w-full {{ classNames ? classNames : '' }} "
        >
            <app-icon matPrefix class="-ml-3 mr-1">search</app-icon>
            <input
                matInput
                #searchInput
                autocomplete="off"
                type="text"
                [placeholder]="placeholder"
                [(ngModel)]="search"
            />
            <button
                *ngIf="searchInput.value?.length"
                class="sm text-xs"
                matSuffix
                icon
                matRipple
                aria-label="Clear"
                (click)="clearSearch()"
            >
                <app-icon>close</app-icon>
            </button>
        </mat-form-field>
    `,
})
export class SearchInputComponent
    extends AsyncHandler
    implements AfterViewInit
{
    @Input() placeholder: string = '';
    @Input() classNames: string = '';
    @Input() search: string = '';
    @Output() onSearch: EventEmitter<string> = new EventEmitter<string>();
    @ViewChild('searchInput', { static: true }) input: ElementRef;

    search$: Observable<string>;

    constructor() {
        super();
    }

    clearSearch() {
        this.input.nativeElement.value = '';
        this.onSearch.emit('');
    }

    ngAfterViewInit() {
        this.search$ = fromEvent<any>(this.input.nativeElement, 'keyup').pipe(
            map((event) => event.target.value),
            debounceTime(500),
            distinctUntilChanged()
        );
        this.subscription(
            'search-input',
            this.search$.subscribe((value) => this.onSearch.emit(value))
        );
    }
}
