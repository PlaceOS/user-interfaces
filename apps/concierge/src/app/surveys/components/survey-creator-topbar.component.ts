import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { AsyncHandler } from '@placeos/common';
import { OrganisationService } from '@placeos/organisation';
import { TriggerOptions } from '@placeos/survey-suite';
import { combineLatest } from 'rxjs';
import { map, shareReplay } from 'rxjs/operators';
import { SurveyOptions, SurveyService } from '../services/survey.service';

@Component({
    selector: 'survey-creator-topbar',
    styles: [],
    template: `
        <div
            class="flex flex-row items-center justify-between bg-base-100 px-4 w-full space-x-2"
        >
            <div
                class="flex flex-1 space-x-2 pt-2 w-[calc(100%-18rem)] max-w-[44rem] overflow-hidden"
            >
                <div class="flex-1 w-1/4">
                    <label>Building</label>
                    <mat-form-field
                        appearance="outline"
                        class="w-full no-subscript"
                    >
                        <mat-select
                            [value]="(options$ | async).building_id"
                            (valueChange)="onBuildingChange($event)"
                        >
                            <mat-option
                                *ngFor="let b of buildings$ | async"
                                [value]="b.id"
                                >{{ b.display_name || b.name }}</mat-option
                            >
                        </mat-select>
                    </mat-form-field>
                </div>

                <div class="flex-1 w-1/4">
                    <label>Level</label>
                    <mat-form-field
                        appearance="outline"
                        class="w-full no-subscript"
                    >
                        <mat-select
                            [value]="(options$ | async).zone_id"
                            (valueChange)="updateOptions({ zone_id: $event })"
                        >
                            <mat-option
                                [value]="(options$ | async).building_id"
                            >
                                All Levels
                            </mat-option>
                            <mat-option
                                *ngFor="let level of levels$ | async"
                                [value]="level.id"
                            >
                                {{ level.display_name || level.name }}
                            </mat-option>
                        </mat-select>
                    </mat-form-field>
                </div>

                <div class="flex-1 w-1/4">
                    <label>Trigger</label>
                    <mat-form-field
                        appearance="outline"
                        class="w-full no-subscript"
                    >
                        <mat-select
                            [value]="(options$ | async).trigger"
                            (valueChange)="updateOptions({ trigger: $event })"
                        >
                            <mat-option
                                *ngFor="let op of triggerOptions"
                                [value]="op.value"
                            >
                                {{ op.name }}
                            </mat-option>
                        </mat-select>
                    </mat-form-field>
                </div>
            </div>
            <div class="flex items-center ml-auto space-x-2 mt-8">
                <button btn matRipple class="inverse w-32" (click)="onCancel()">
                    Cancel
                </button>
                <button btn matRipple class="w-32" (click)="onSave()">
                    Save
                </button>
            </div>
        </div>
    `,
})
export class SurveyCreatorTopbarComponent
    extends AsyncHandler
    implements OnInit
{
    public selected_building = null;
    public levels = [];
    public readonly options$ = this._survey._options$.pipe(shareReplay(1));
    public readonly buildings$ = this._org.building_list;
    public readonly levels$ = combineLatest([
        this.options$,
        this._org.level_list,
    ]).pipe(
        map(([options, levels]) => {
            const { building_id } = options;
            if (!building_id?.length) return [];
            return levels.filter((l) => l.parent_id === building_id);
        })
    );
    public readonly triggerOptions = TriggerOptions;

    updateOptions = (op: Partial<SurveyOptions>) => this._survey.setOptions(op);
    onSave = () => this._survey.saveSurvey();
    onCancel = () => this._survey.back();

    constructor(
        private _org: OrganisationService,
        private _actRoute: ActivatedRoute,
        private _survey: SurveyService
    ) {
        super();
    }

    async ngOnInit() {
        this.subscription(
            'query_params',
            this._actRoute.queryParams.subscribe((qparams) => {
                const id = qparams['building_id'];
                if (id?.length) this.updateOptions({ building_id: id });
            })
        );
    }

    onBuildingChange(building_id: string) {
        if (!building_id?.length) return;
        this.doBuildingChange(building_id);
    }

    private doBuildingChange(building_id: string) {
        if (!building_id?.length) return;
        this.updateOptions({ zone_id: building_id, building_id });
    }
}
