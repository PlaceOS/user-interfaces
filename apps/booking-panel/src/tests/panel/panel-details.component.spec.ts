import { createRoutingFactory, SpectatorRouting } from '@ngneat/spectator/jest';

import { PanelStateService } from '../../app/panel/panel-state.service';
import { PanelDetailsComponent } from '../../app/panel/panel-details.component';
import { BehaviorSubject } from 'rxjs';
import { MockComponent } from 'ng-mocks';
import { IconComponent } from '@placeos/components';

describe('PanelDetailsComponent', () => {
    let spectator: SpectatorRouting<PanelDetailsComponent>;
    const createComponent = createRoutingFactory({
        component: PanelDetailsComponent,
        declarations: [MockComponent(IconComponent)],
        providers: [
            {
                provide: PanelStateService,
                useValue: {
                    space: new BehaviorSubject(null),
                    settings: new BehaviorSubject({}),
                    newBooking: jest.fn(),
                    confirmWaiter: jest.fn(),
                    viewControl: jest.fn(),
                },
            },
        ],
    });

    beforeEach(() => {
        localStorage.setItem('PLACEOS.BOOKINGS.system', 'a-system');
        spectator = createComponent();
    });

    afterEach(() => localStorage.clear());

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
    });

    it('should allow for creating new events', () => {
        const service = spectator.inject(PanelStateService);
        expect('[event]').toExist();
        spectator.click('[event]');
        expect(service.newBooking).toHaveBeenCalled();
        (service.settings as any).next({ disable_book_now: true });
        spectator.detectChanges();
        expect('[event]').not.toExist();
    });

    it('should allow for calling waiter', () => {
        expect('[waiter]').not.toExist();
        const service = spectator.inject(PanelStateService);
        (service.settings as any).next({ catering_ui: true });
        spectator.detectChanges();
        expect('[waiter]').toExist();
        spectator.click('[waiter]');
        expect(service.confirmWaiter).toHaveBeenCalled();
    });

    it('should allow for viewing control UI', () => {
        expect('[control]').not.toExist();
        const service = spectator.inject(PanelStateService);
        (service.settings as any).next({ control_ui: 'here' });
        spectator.detectChanges();
        expect('[control]').toExist();
        spectator.click('[control]');
        expect(service.viewControl).toHaveBeenCalled();
    });
});
