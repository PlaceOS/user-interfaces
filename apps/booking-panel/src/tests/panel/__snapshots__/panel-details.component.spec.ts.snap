// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PanelDetailsComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <h1
    class="text-4xl font-medium mb-8"
  >
     &lt;Unknown Space&gt; 
  </h1>
  
  <div
    class="flex flex-col items-center space-y-2"
  >
    <button
      class="w-48"
      event=""
      mat-button=""
    >
      <div
        class="flex items-center justify-center"
      >
        <app-icon
          class="text-xl"
        >
          add_circle_outline
        </app-icon>
        <span
          class="mx-2"
        >
          New Event
        </span>
      </div>
    </button>
    
    
    
  </div>
  
  
</div>
`;
